controllers/nutanixmachine_controller.go

```diff
@@ -565,60 +566,77 @@ func (r *NutanixMachineReconciler) checkFailureDomainStatus(rctx *nctx.MachineCo
+	// TODO: @faiq -- to handle the legacy failure domains this function checks to see if fdName
+	// is present in the legacy embedded field. if it is, we return a "dummy" spec for the new failure domain
+	// CR with the subnets and cluster info
+	failureDomainName := *rctx.Machine.Spec.FailureDomain
+	if rctx.NutanixCluster != nil && rctx.NutanixCluster.Spec.FailureDomains != nil { //nolint:staticcheck // this handles old field
+		failureDomain, err := GetLegacyFailureDomainFromNutanixCluster(failureDomainName, rctx.NutanixCluster)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find failure domain %s", failureDomainName)
+		}
+		if failureDomain != nil {
+			cluster := failureDomain.Cluster
+			subnets := failureDomain.Subnets
+			fdSpec := &infrav1.NutanixFailureDomainSpec{
+				PrismElementCluster: cluster,
+				Subnets:             subnets,
+			}
+			return fdSpec, nil
+		}
+	}
+	// if the old field wasn't set or the failure domain name referenced isn't present there, we
+	// can assume that it is refering to the new CRD so we make a get
 	fdObj := &infrav1.NutanixFailureDomain{}
 	fdKey := client.ObjectKey{Name: fdName, Namespace: rctx.NutanixMachine.Namespace}
 	if err := r.Get(rctx.Context, fdKey, fdObj); err != nil {
 		return nil, fmt.Errorf("failed to fetch the referent failure domain object %q: %w", fdName, err)
 	}
-	return fdObj, nil
+	return &fdObj.Spec, nil
 }
 
-func (r *NutanixMachineReconciler) validateFailureDomainRef(rctx *nctx.MachineContext, fdName string) (*infrav1.NutanixFailureDomain, error) {
-	// Fetch the referent failure domain object
-	fdObj, err := r.getFailureDomainObj(rctx, fdName)
-	if err != nil {
-		return nil, err
-	}
-
+func (r *NutanixMachineReconciler) validateFailureDomainSpec(rctx *nctx.MachineContext, fdSpec *infrav1.NutanixFailureDomainSpec) error {
 	// Validate the failure domain configuration
-	pe := fdObj.Spec.PrismElementCluster
+	pe := fdSpec.PrismElementCluster
 	peUUID, err := GetPEUUID(rctx.Context, rctx.NutanixClient, pe.Name, pe.UUID)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	subnets := fdObj.Spec.Subnets
+	subnets := fdSpec.Subnets
 	_, err = GetSubnetUUIDList(rctx.Context, rctx.NutanixClient, subnets, peUUID)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return fdObj, nil
+	return nil
 }
 
 func (r *NutanixMachineReconciler) validateMachineConfig(rctx *nctx.MachineContext) error {
 	log := ctrl.LoggerFrom(rctx.Context)
-
 	fdName := rctx.Machine.Spec.FailureDomain
 	if fdName != nil && *fdName != "" {
 		log.WithValues("failureDomain", *fdName)
-		fdObj, err := r.validateFailureDomainRef(rctx, *fdName)
+		fdSpec, err := r.getFailureDomainSpec(rctx, *fdName)
 		if err != nil {
-			log.Error(err, "Failed to validate the failure domain")
+			log.Error(err, fmt.Sprintf("Failed to get the failure domain object %s", *fdName))
```

Comment:
remove "object" in the error message, since in the legacy case the FD is not object.